storage/src/main/java/org/apache/kafka/storage/internals/log/VerificationGuard.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.storage.internals.log;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class VerificationGuard {
+
+    // The sentinel VerificationGuard will be used as a default when no verification guard is provided.
+    // It can not be used to verify a transaction is ongoing and its value is always 0.
+    public static final VerificationGuard SENTINEL = new VerificationGuard(0);
+    private static final AtomicLong INCREMENTING_ID = new AtomicLong(0L);
+    private final long value;
+
+    public VerificationGuard() {
+        value = INCREMENTING_ID.incrementAndGet();
+    }
+
+    private VerificationGuard(long value) {
+        this.value = value;
+    }
+
+    @Override
+    public String toString() {
+        return "VerificationGuard: " + value;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if ((null == obj) || (obj.getClass() != this.getClass()))
+            return false;
+        VerificationGuard guard = (VerificationGuard) obj;
+        return value == guard.value();
+    }
+
+    @Override
+    public int hashCode() {
+        return Long.hashCode(value);
+    }
+
+    private long value() {
+        return value;
+    }
+
+    public boolean verifiedBy(VerificationGuard verifyingGuard) {
```

Comment:
Nit: would `verify` be a better method name? It's the active voice which is generally better for methods.

Reply:
I didn't know if verify sounded like it would return a boolean. But I suppose it can.

Reply:
Another active voice option is "verifies" (kind of like "equals").

Reply:
Interesting, what would you expect `verify` to return? Typically, a given method would return a value that gives you the result for the operation you performed. In this case, it can be a boolean if it's a `true`/`false` case or something more sophisticated if there are more states.

Reply:
I guess it sounded more like a void/unit type method that does an action. But yeah, then I guess the result is true/false if the action had a conclusion. I can change to verify if we think that is best.

Reply:
I guess I also though that it was weird to be an instance method if it is verify. (Verify could be a static method that took two arguments.)

Reply:
How is `verify` different than `map`, `filter` and other similar instance methods? I am asking these questions because naming conventions are important overall, not due to this one example.

Reply:
Map and filter return types that are clearer to their names ie map is usually another structure of the same or a new type and filter is usually the same type. 

I guess when you verify something the output is either "yes this was verified or no it is not". So I can see how you could get a boolean there. I could also see verification as an action that is silent and throws an error if it failed. 

Maybe I'm just being too grammatically minded here.  I would say "x verifies y" and not "x verify y". For map and filter I see it as "map x to z using y" and "filter x to z using y" where x is the instance, y is the argument, and z is the output. For verify it would read as "Verify x using y."  It's not a huge deal though so I can just change to verify.